### PR TITLE
Package basics

### PR DESCRIPTION
**Pull request by theopolis**
I'm changing how make_osx_package.sh works again, sorry. But now `make package` will build us a PKG that plays nice with existing installs and for deployments managing their own osquery.conf/com.facebook.osquery.plist (LD). 

I've added liberal package dependencies for Ubuntu and tested with trusty. I've also improved compile times for tests by requiring an explicit argument to link against the extended (additional) set of tables for test targets.

### CODE REVIEW

**Comment by theopolis** [comment]
@javuto, @marpaia I'm thinking running `make package` after each commit to master. Then shipping that output package to a signing server or to a svn repo hosted alongside the public website or LP would be a good first step for packages.

File: osquery/CMakeLists.txt [hunk 1]
```diff
@@ -113,44 +114,66 @@ ADD_OSQUERY_LIBRARY(osquery_amalgamation ${AMALGAMATION})
...
+  install(PROGRAMS "${CMAKE_SOURCE_DIR}/tools/deployment/osquery.initd"
+    DESTINATION "/etc/init.d/" RENAME "osquery" COMPONENT main)
+endif()
+
+# make package
+if(APPLE)
+  cmake_policy(SET CMP0037 OLD)
+  add_custom_target(
+    package
+    ./tools/deployment/make_osx_package.sh
```
Comment (marpaia) [inline osquery/CMakeLists.txt]: do we want to build with -n by default? or do we want to make -n the default if that's the case?

File: osquery/CMakeLists.txt [hunk 2]
```diff
@@ -113,44 +114,66 @@ ADD_OSQUERY_LIBRARY(osquery_amalgamation ${AMALGAMATION})
...
+    WORKING_DIRECTORY "${CMAKE_SOURCE_DIR}"
+    COMMENT "Building default OS X package (no custom config)" VERBATIM
+  )
+elseif(LINUX)
   set(CPACK_PACKAGE_NAME "osquery")
   set(CPACK_PACKAGE_VENDOR "facebook")
+  set(CPACK_PACKAGE_CONTACT "osquery@osquery.io")
```
Comment (marpaia) [inline osquery/CMakeLists.txt]: does this exist?

File: osquery/tables/CMakeLists.txt [hunk 3]
```diff
@@ -107,10 +107,10 @@ ADD_OSQUERY_CORE_LIBRARY(osquery_utility_tables
...
 )
 
-ADD_OSQUERY_TEST(etc_hosts_tests networking/etc_hosts_tests.cpp)
+ADD_OSQUERY_TEST(etc_hosts_tests networking/etc_hosts_tests.cpp TRUE)
 if(APPLE)
-  ADD_OSQUERY_TEST(apps_tests system/darwin/apps_tests.cpp)
-  ADD_OSQUERY_TEST(ca_certs_tests system/darwin/ca_certs_tests.cpp)
-  ADD_OSQUERY_TEST(firewall_tests system/darwin/firewall_tests.cpp)
-  ADD_OSQUERY_TEST(launchd_tests system/darwin/launchd_tests.cpp)
+  ADD_OSQUERY_TEST(apps_tests system/darwin/apps_tests.cpp TRUE)
```
Comment (marpaia) [inline osquery/tables/CMakeLists.txt]: I tried looking through the code for ADD_OSQUERY_TEST, but I can't really figure out what the TRUE is for here.

File: tools/deployment/make_osx_package.sh [hunk 4]
```diff
@@ -46,29 +44,39 @@ set -e
...
+    -l PATH override the default launchd plist.
+    -o PATH override the output path.
+
+  This will generate an OSX package with:
+  (1) An example config /var/osquery/osquery.example.config
+  (2) An optional config /var/osquery/osquery.config if [-c] is used
+  (3) A LaunchDaemon plist /var/osquery/com.facebook.osqueryd.plist
+  (4) The osquery toolset /usr/local/bin/osquery*
+
+  To enable osqueryd to run at boot using Launchd use osqueryctl.
```
Comment (marpaia) [inline tools/deployment/make_osx_package.sh]: should we add osqueryctl before we add this? or should we just make sure it's there before we cut the next release?

File: osquery/CMakeLists.txt (hunk 1 above)
Comment (theopolis) [inline osquery/CMakeLists.txt]: I took out -n already. ;)

File: osquery/CMakeLists.txt (hunk 2 above)
Comment (theopolis) [inline osquery/CMakeLists.txt]: It should, or we can maintain something FB-managed. I'm not opposed to an internal list as long as it's easy to add external addresses.

File: osquery/tables/CMakeLists.txt (hunk 3 above)
Comment (theopolis) [inline osquery/tables/CMakeLists.txt]: This is a gotcha for supporting optional arguments in macros. The TRUE means link both basic and additional libosquery.

File: tools/deployment/make_osx_package.sh (hunk 4 above)
Comment (theopolis) [inline tools/deployment/make_osx_package.sh]: I'd vote for having osqueryctl in before the release cut.

**Comment by theopolis** [comment]
More work ahead:
```
[vagrant@localhost vagrant]$ ldd ./build/linux/osquery/osqueryi 
	linux-vdso.so.1 =>  (0x00007fff10dff000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fdd58bcc000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fdd589b5000)
	libsnappy.so.1 => /usr/lib64/libsnappy.so.1 (0x00007fdd587b0000)
	libcrypto.so.10 => /usr/lib64/libcrypto.so.10 (0x00007fdd583cd000)
	libssl.so.10 => /usr/lib64/libssl.so.10 (0x00007fdd58160000)
	libreadline.so.6 => /lib64/libreadline.so.6 (0x00007fdd57f1d000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fdd57d19000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fdd57b07000)
	liblzma.so.0 => /usr/lib64/liblzma.so.0 (0x00007fdd578e6000)
	libunwind.so.8 => /usr/lib64/libunwind.so.8 (0x00007fdd576cd000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fdd574c4000)
	libboost_program_options.so.1.55.0 => /usr/local/lib/libboost_program_options.so.1.55.0 (0x00007fdd57258000)
	librpm.so.1 => /usr/lib64/librpm.so.1 (0x00007fdd56fed000)
	librpmio.so.1 => /usr/lib64/librpmio.so.1 (0x00007fdd56dbd000)
	libproc-3.2.8.so => /lib64/libproc-3.2.8.so (0x00007fdd56b9a000)
	libudev.so.0 => /lib64/libudev.so.0 (0x00007fdd5698c000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007fdd56685000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fdd56401000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fdd561eb000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fdd55e56000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fdd58df1000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007fdd55c12000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007fdd5592c000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007fdd55727000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007fdd554fb000)
	libtinfo.so.5 => /lib64/libtinfo.so.5 (0x00007fdd552da000)
	libnss3.so => /usr/lib64/libnss3.so (0x00007fdd54f9a000)
	libelf.so.1 => /usr/lib64/libelf.so.1 (0x00007fdd54d84000)
	liblua-5.1.so => /usr/lib64/liblua-5.1.so (0x00007fdd54b57000)
	libpopt.so.0 => /lib64/libpopt.so.0 (0x00007fdd5494d000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fdd5472e000)
	libcap.so.2 => /lib64/libcap.so.2 (0x00007fdd5452a000)
	libacl.so.1 => /lib64/libacl.so.1 (0x00007fdd54321000)
	libdb-4.7.so => /lib64/libdb-4.7.so (0x00007fdd53fad000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007fdd53da1000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007fdd53b9e000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fdd53984000)
	libnssutil3.so => /usr/lib64/libnssutil3.so (0x00007fdd53757000)
	libplc4.so => /lib64/libplc4.so (0x00007fdd53552000)
	libplds4.so => /lib64/libplds4.so (0x00007fdd5334e000)
	libnspr4.so => /lib64/libnspr4.so (0x00007fdd5310f000)
	libattr.so.1 => /lib64/libattr.so.1 (0x00007fdd52f0a000)
```

**Comment by theopolis** [comment]
Without program_options:
```
agrant@localhost vagrant]$ ldd ./build/linux/osquery/osqueryi 
	linux-vdso.so.1 =>  (0x00007fff9f1ff000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f16dea85000)
	libz.so.1 => /lib64/libz.so.1 (0x00007f16de86e000)
	libsnappy.so.1 => /usr/lib64/libsnappy.so.1 (0x00007f16de669000)
	libcrypto.so.10 => /usr/lib64/libcrypto.so.10 (0x00007f16de286000)
	libssl.so.10 => /usr/lib64/libssl.so.10 (0x00007f16de019000)
	libreadline.so.6 => /lib64/libreadline.so.6 (0x00007f16dddd6000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f16ddbd2000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007f16dd9c0000)
	liblzma.so.0 => /usr/lib64/liblzma.so.0 (0x00007f16dd79f000)
	libunwind.so.8 => /usr/lib64/libunwind.so.8 (0x00007f16dd586000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f16dd37d000)
	librpm.so.1 => /usr/lib64/librpm.so.1 (0x00007f16dd112000)
	librpmio.so.1 => /usr/lib64/librpmio.so.1 (0x00007f16dcee3000)
	libproc-3.2.8.so => /lib64/libproc-3.2.8.so (0x00007f16dccbf000)
	libudev.so.0 => /lib64/libudev.so.0 (0x00007f16dcab1000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f16dc7ab000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f16dc526000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f16dc310000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f16dbf7c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f16decaa000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007f16dbd37000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007f16dba51000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007f16db84d000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007f16db620000)
	libtinfo.so.5 => /lib64/libtinfo.so.5 (0x00007f16db3ff000)
	libnss3.so => /usr/lib64/libnss3.so (0x00007f16db0bf000)
	libelf.so.1 => /usr/lib64/libelf.so.1 (0x00007f16daea9000)
	liblua-5.1.so => /usr/lib64/liblua-5.1.so (0x00007f16dac7c000)
	libpopt.so.0 => /lib64/libpopt.so.0 (0x00007f16daa72000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007f16da853000)
	libcap.so.2 => /lib64/libcap.so.2 (0x00007f16da64f000)
	libacl.so.1 => /lib64/libacl.so.1 (0x00007f16da446000)
	libdb-4.7.so => /lib64/libdb-4.7.so (0x00007f16da0d2000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007f16d9ec6000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007f16d9cc3000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007f16d9aa9000)
	libnssutil3.so => /usr/lib64/libnssutil3.so (0x00007f16d987c000)
	libplc4.so => /lib64/libplc4.so (0x00007f16d9677000)
	libplds4.so => /lib64/libplds4.so (0x00007f16d9473000)
	libnspr4.so => /lib64/libnspr4.so (0x00007f16d9234000)
	libattr.so.1 => /lib64/libattr.so.1 (0x00007f16d902f000)
```

**Comment by theopolis** [comment]
Without rpm/rpmio:
```
[vagrant@localhost vagrant]$ ldd ./build/linux/osquery/osqueryi 
	linux-vdso.so.1 =>  (0x00007fffca3ff000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fa818c91000)
	libz.so.1 => /lib64/libz.so.1 (0x00007fa818a7a000)
	libsnappy.so.1 => /usr/lib64/libsnappy.so.1 (0x00007fa818875000)
	libcrypto.so.10 => /usr/lib64/libcrypto.so.10 (0x00007fa818492000)
	libssl.so.10 => /usr/lib64/libssl.so.10 (0x00007fa818225000)
	libreadline.so.6 => /lib64/libreadline.so.6 (0x00007fa817fe2000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fa817dde000)
	libbz2.so.1 => /lib64/libbz2.so.1 (0x00007fa817bcc000)
	liblzma.so.0 => /usr/lib64/liblzma.so.0 (0x00007fa8179ab000)
	libunwind.so.8 => /usr/lib64/libunwind.so.8 (0x00007fa817792000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fa817589000)
	libproc-3.2.8.so => /lib64/libproc-3.2.8.so (0x00007fa817366000)
	libudev.so.0 => /lib64/libudev.so.0 (0x00007fa817158000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007fa816e51000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fa816bcd000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fa8169b7000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fa816622000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fa818eb6000)
	libgssapi_krb5.so.2 => /lib64/libgssapi_krb5.so.2 (0x00007fa8163de000)
	libkrb5.so.3 => /lib64/libkrb5.so.3 (0x00007fa8160f8000)
	libcom_err.so.2 => /lib64/libcom_err.so.2 (0x00007fa815ef3000)
	libk5crypto.so.3 => /lib64/libk5crypto.so.3 (0x00007fa815cc7000)
	libtinfo.so.5 => /lib64/libtinfo.so.5 (0x00007fa815aa6000)
	libkrb5support.so.0 => /lib64/libkrb5support.so.0 (0x00007fa81589a000)
	libkeyutils.so.1 => /lib64/libkeyutils.so.1 (0x00007fa815697000)
	libresolv.so.2 => /lib64/libresolv.so.2 (0x00007fa81547d000)
	libselinux.so.1 => /lib64/libselinux.so.1 (0x00007fa81525d000)
```

Remaining concerns: snappy. Differences from linux: (1) ssl/crypto are pulling in a few more dependencies, (2) rt and proc are only available as shared libraries.

**Pull request merged**